Add custom task tutorial (#1057)
deepchecks/deepchecks: docs/source/tutorials/vision/custom_task_tutorial.rst

====================

> Custom Task Tutorial

Reviewer: pretty long, needs table of contents?

====================

> Computer vision is consists of different objectives models are trained for. The desired objective reflects

Reviewer: ```suggestion
Computer vision is an umbrella term for a wide spectrum of objectives models are trained for. These objective reflect
```

on the structure of the data and the possible actions on it.

> In Deepchecks the first step before running checks is to create an implementation of `VisionData` which

Reviewer: ```suggestion
The first step before running any Deepchecks checks is to create an implementation of `VisionData`, which
```

> standardize the structure of the data for a given computer vision task.

Reviewer: link to tutorials

> Some of the checks are agnostic to the specific task and will work on all of them, and some are tied to the

Reviewer: This part should be improved - start by saying what current tasks are supported, then give example for checks that can run for other tasks and explain why is that (the last part is what you explained here).

specific details of a task and won't work for other tasks. When implementing your own task all of the
task-agnostic checks will be able to run.

Defining the Data & Model
=========================

> First we will define a `Dataset` of COCO-128 segmentation task.

Reviewer: maybe note that this is a setup stage and that the deepchecks specific code will be in the next section?

          for i in range(5)]

  show(imgs)

.. image :: /_static/custom-segment-tutorial.png
  :width: 400
  :alt: COCO128 segmentation visualize

For the model we will use a pretrain model from `torchvision`

.. code-block:: python

  from torchvision.models.detection import fasterrcnn_resnet50_fpn

  model = fasterrcnn_resnet50_fpn(pretrained=True, progress=False)
  model = model.eval()

Implement Custom `VisionData`
=============================

When we have our `DataLoader`s and model ready we can start creating our own implementation
of `VisionData` class.

.. code-block:: python

  from typing import List, Iterable
  from deepchecks.vision.vision_data import VisionData

  class MyCustomSegmentationData(VisionData):
      """Class for loading the COCO segmentation dataset."""

      def get_classes(self, batch_labels) -> List[List[int]]:
          """Return per label a list of classes (by id) in it."""
          # The input `batch_labels` is the result of `batch_to_labels` function.
          return batch_labels[0]

      def batch_to_labels(self, batch):
          """Extract from the batch only the labels."""

>           return batch[1], batch[2]

Reviewer: explain what happens here

      def infer_on_batch(self, batch, model, device):
          """Infer on a batch of images."""
          predictions = model.to(device)(batch[0])
          return predictions

      def batch_to_images(self, batch) -> Iterable[np.ndarray]:
          """Convert the batch to a list of images as (H, W, C) 3D numpy array per image."""
          return [tensor.numpy().transpose((1, 2, 0)) for tensor in batch[0]]

> Now we are able to run checks that are working only on the image data, since it's in the standard format of

Reviewer: ```suggestion
Now we are able to run checks that use only the image data, since it's in the standard Deepchecks format
```
Link to explanation of image format.

> Deepchecks. For more checks we'll need to define custom properties or metrics.

Reviewer: ```suggestion
Deepchecks. In order for additional check types to run, we'll need to define custom properties or metrics.
```

Implement Custom Properties
===========================

In order to run checks that are using label or prediction properties we we'll have to implement
a custom property. For a more in-depth explanation on properties see
:doc:`Data Properties </user-guide/vision/vision_properties>`

.. code-block:: python

  #### Label properties

>   def number_of_detections(labels) -> List[int]:

Reviewer: would have been cooler to use area

Author: Maybe, but it's harder and not the focus of the tutorial

  """Return a list containing the number of detections per sample in batch."""
  # Labels object is the result of `batch_to_labels` function
  classes, all_masks = labels
  return [sample_masks.shape[0] for sample_masks in all_masks]

  def classes_in_labels(labels: List[torch.Tensor]) -> List[int]:
    """Return a list containing the classes in batch."""
    # Labels object is the result of `batch_to_labels` function
    classes, all_masks = labels
    return list(chain.from_iterable(classes))

  # We will pass this object as parameter to checks that are using label properties
  label_properties = [
    {'name': '# Detections per Label', 'method': number_of_detections, 'output_type': 'discrete'},
    {'name': 'Classes in Labels', 'method': classes_in_labels, 'output_type': 'class_id'}
  ]

Implement Custom Metric
=======================

> Some checks requires a metric in order to run. When using a custom task you will also have to create a custom

Reviewer: ```suggestion
Some checks test the model performance and requires a metric in order to run. When using a custom task you will also have to create a custom
```

> metric in order for those checks to work, since the built-in metrics does not know to handle your data structure.

Reviewer: ```suggestion
metric in order for those checks to work, since the built-in metrics don't know to handle your data structure.
```
